crates/noirc_abi/src/input_parser/json.rs

```diff
@@ -0,0 +1,152 @@
+use super::{parse_str_to_field, InputValue};
+use crate::{errors::InputParserError, Abi, AbiType, MAIN_RETURN_NAME};
+use acvm::FieldElement;
+use iter_extended::{btree_map, try_btree_map, try_vecmap, vecmap};
+use serde::{Deserialize, Serialize};
+use std::collections::BTreeMap;
+
+pub(crate) fn parse_json(
+    input_string: &str,
+    abi: &Abi,
+) -> Result<BTreeMap<String, InputValue>, InputParserError> {
+    // Parse input.json into a BTreeMap.
+    let data: BTreeMap<String, JsonTypes> = serde_json::from_str(input_string)?;
+
+    // Convert arguments to field elements.
+    let mut parsed_inputs = try_btree_map(abi.to_btree_map(), |(arg_name, abi_type)| {
+        // Check that json contains a value for each argument in the ABI.
+        let value = data
+            .get(&arg_name)
+            .ok_or_else(|| InputParserError::MissingArgument(arg_name.clone()))?;
+
+        InputValue::try_from_json(value.clone(), &abi_type, &arg_name)
+            .map(|input_value| (arg_name, input_value))
+    })?;
+
+    // If the json file also includes a return value then we parse it as well.
+    // This isn't required as the prover calculates the return value itself.
+    if let (Some(return_type), Some(json_return_value)) =
+        (&abi.return_type, data.get(MAIN_RETURN_NAME))
+    {
+        let return_value =
+            InputValue::try_from_json(json_return_value.clone(), return_type, MAIN_RETURN_NAME)?;
+        parsed_inputs.insert(MAIN_RETURN_NAME.to_owned(), return_value);
+    }
+
+    Ok(parsed_inputs)
+}
+
+pub(crate) fn serialize_to_json(
+    w_map: &BTreeMap<String, InputValue>,
+) -> Result<String, InputParserError> {
+    let to_map: BTreeMap<_, _> =
+        w_map.iter().map(|(key, value)| (key, JsonTypes::from(value.clone()))).collect();
+
+    let json_string = serde_json::to_string(&to_map)?;
+
+    Ok(json_string)
+}
+
+#[derive(Debug, Deserialize, Serialize, Clone)]
+#[serde(untagged)]
+enum JsonTypes {
+    // This is most likely going to be a hex string
+    // But it is possible to support UTF-8
+    String(String),
+    // Just a regular integer, that can fit in 128 bits
```

Comment:
should this be a u128, since comments say it fits in 128 bit?

Reply:
No, json and toml can't represent u128s. In fact javascript only supports ~2^53 so I'll do some bounds checking to see if we can tighten this up.

Reply:
Then the comment is wrong no?

> Just a regular integer, that can fit in 128 bits

Thats a u128.

> In fact javascript only supports ~2^53 so I'll do some bounds checking to see if we can tighten this up

I think we want this to be the same for both TOML and JSON, ie not u64 for TOML and perhaps u52 or u32 for JSON.

Changing the TOML would be a breaking change though so that would need to be done in a separate PR if we were to do that

Reply:
Makes sense - I think this is the same for TOML too; ie there is no specific integer max, see : https://github.com/toml-lang/toml/issues/688

Reply:
Sort of, the spec calls out i64s specifically https://toml.io/en/v1.0.0#integer. Granted this is a minimum but toml-rs only supports this for compatibility reasons.